op-challenger/game/fault/trace/cannon/provider_test.go

```diff
@@ -215,14 +246,37 @@ func setupTestData(t *testing.T) (string, string) {
 	return dataDir, "state.json"
 }
 
-func setupWithTestData(t *testing.T, dataDir string, prestate string) (*CannonTraceProvider, *stubGenerator) {
+func setupWithTestData(t *testing.T, dataDir string, prestate string) (*CannonTraceProvider, *stubGenerator, *stubStepper) {
 	generator := &stubGenerator{}
+	stepper := &stubStepper{}
 	return &CannonTraceProvider{
-		logger:    testlog.Logger(t, log.LvlInfo),
-		dir:       dataDir,
-		generator: generator,
-		prestate:  filepath.Join(dataDir, prestate),
-	}, generator
+		logger:         testlog.Logger(t, log.LvlInfo),
+		dir:            dataDir,
+		generator:      generator,
+		prestate:       filepath.Join(dataDir, prestate),
+		lastStepReader: stepper.loadLastStep,
+		lastStepWriter: stepper.storeLastStep,
+	}, generator, stepper
+}
+
+type stubStepper struct {
```

Comment:
I don't think I'd use the stub stepper here.  The other tests in this file read and write to real disk already because the majority of the code in the provider is disk access.  I think the production code would be slightly simpler without supporting a stub and then we'd actually test the read/write code for the final state as well which is currently missed.

The test would basically be to create a trace provider and request data after the end of the trace - it should then write the final state file.  Then create a new trace provider from the same dir and it should read that state file without having to re-generate.